Tests/MQTTnet.Core.Tests/MQTTv5/ExtendedAuth/InteractiveEnhancedAuthBrokerHandler.cs

```diff
@@ -0,0 +1,61 @@
+using System.Collections.Generic;
+using System.Text;
+using MQTTnet.Packets;
+using MQTTnet.Protocol;
+using MQTTnet.Server.ExtendedAuthenticationExchange;
+
+namespace MQTTnet.Tests.MQTTv5.ExtendedAuth
+{
+    public class InteractiveEnhancedAuthBrokerHandler : IMqttEnhancedAuthenticationBrokerHandler
+    {
+        public MqttBasePacket HandleAuth(MqttAuthPacket authPacketUpdate,
+            IDictionary<object, object> sessionItems)
+        {
+            if (authPacketUpdate.Properties.AuthenticationMethod != AuthMethod.InteractiveAuthName)
+            {
+                return null;
+            }
+
+            var secret = Encoding.UTF8.GetString(authPacketUpdate.Properties.AuthenticationData);
+            if (secret == "nonce2_12345")
+            {
+                return new MqttConnAckPacket
+                {
+                    ReturnCode = MqttConnectReturnCode.ConnectionAccepted,
+                    ReasonCode = MqttConnectReasonCode.Success,
+                    Properties = new MqttConnAckPacketProperties
+                    {
+                        AuthenticationMethod = AuthMethod.InteractiveAuthName
+                    }
+                };
+            }
+
+            return new MqttAuthPacket
+            {
+                ReasonCode = MqttAuthenticateReasonCode.ContinueAuthentication,
+                Properties = new MqttAuthPacketProperties
+                {
+                    AuthenticationMethod = AuthMethod.InteractiveAuthName,
+                    AuthenticationData = Encoding.UTF8.GetBytes("nonce2")
+                }
+            };
+        }
+
+        public MqttBasePacket StartChallenge(MqttConnectPacket connectPacket)
+        {
+            if (connectPacket.Properties.AuthenticationMethod != AuthMethod.InteractiveAuthName)
```

Comment:
From a pure programming point of view, the fact that I cannot setup the CONNACK packet in this function with all of the failure information is unfortunate.  It would be nice to set errors such as BadAuthenticationMethod into the packet even if I then have the same packet coming into the Connection Validator function.  At that point I can just check if the reason code is not success and return immediately.

Reply:
> From a pure programming point of view, the fact that I cannot setup the CONNACK packet in this function with all of the failure information is unfortunate.

I probably don't understand your point. The StartChallenge method returns just a MqttBasePacket, which means you can setup whatever you like. But setting up a CONNACK packet in this method would be a little unfortunate because, the whole point of this authentication flow is to initiate an enhanced auth.

> It would be nice to set errors such as BadAuthenticationMethod into the packet even if I then have the same packet coming into the Connection Validator function.

It was an intended decision, that if you return null the default implementation kicks in. If you don't want this behavior just return your custom tailored packet, which can be a CONNACK packet as well.

Reply:
I wrote my comment to soon.  I was doing it based on just reading the code rather than having done the integration to my code base and then debugging my process.  Sorry about that.

I am going to change this to a completely different request.  There is now a new item which was defined call SessionItems passing this object into the two functions allows for a clean way to pass data between the call to the connect validator and these two functions.  It also allows for internal state to be kept to deal with multiple round trips such as a Kerberos authentication.